### PR TITLE
soc: stm32l1 power management (stop mode)

### PR DESCRIPTION
**Pull request by abelNoelleC**
Support for STM32L1x power management (right now only stop mode, where CPU is stopped but context is kept).

Requires #32128: 
Reconfiguration of clock required after wake-up. This PR will be updated if #32128 has been merged. 

Next steps:
- low-power timer driver that allows for continuation of tasks post-wake up, for STM32 SoCs that don't have LPTIM - currently in progress: STM32(L1x) RTC timer driver (see #30790 for an idea, further development is currently being done)

Signed-off by: Noelle Clement <n.clement@abelsensors.com>

### CODE REVIEW

File: soc/arm/st_stm32/stm32l1/power.c [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		 */
+
+		/* Enable ultra low power mode */
+		LL_PWR_EnableUltraLowPower();
+
+		/** Set the regulator to low power before setting MODE_STOP.
+		 * If the regulator remains in "main mode",
+		 * it consumes more power without providing any additional feature.
+		 */
+		LL_PWR_SetRegulModeLP(LL_PWR_REGU_LPMODES_LOW_POWER);
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l1/power.c:55]: I can see you don't call `LL_PWR_SetRegulModeLP(LL_PWR_REGU_LPMODES_MAIN);`
Is that done on purpose ?

Comment (abelNoelleC) [reply soc/arm/st_stm32/stm32l1/power.c:55]: Yes, since this decreases the consumption further. Is there anything specific against having this in the STM32L1x-wide sleep mode selection?

Comment (erwango) [reply soc/arm/st_stm32/stm32l1/power.c:55]: My question wasn't clear enough:
I can see you don't call `LL_PWR_SetRegulModeLP(LL_PWR_REGU_LPMODES_MAIN);` **when exiting sleep mode**.
Is that done on purpose ?`

Comment (abelNoelleC) [reply soc/arm/st_stm32/stm32l1/power.c:55]: When I check the reference manual [[here](https://www.st.com/resource/en/reference_manual/cd00240193-stm32l100xx-stm32l151xx-stm32l152xx-and-stm32l162xx-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf)] pg. 122, I see that the voltage regulator is only put in low-power mode during deepsleep/sleep/low-power run mode. If I understand correctly it shouldn't impact the voltage regulator during regular run (so after exiting sleep mode), since it also states "_The regulator goes back to main mode when the CPU exits from these modes_". Was that indeed your concern?

Comment (erwango) [reply soc/arm/st_stm32/stm32l1/power.c:55]: Ok. Indeed that's a good thing to set this.
Though, I'd rather move this in `stm32_power_init` as current function is a transition function. We expect to see here things that are need to be set when entering sleep mode and because they are unset when exiting sleep mode.

Comment (erwango) [inline soc/arm/st_stm32/stm32l1/power.c:55]: Please move these 2 functions in `stm32_power_init`

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by erwango** [comment]
Setting DNM as this requires https://github.com/zephyrproject-rtos/zephyr/pull/30790 to be available before being merged. @abelNoelleC Do you confirm this is correct understanding?

File: soc/arm/st_stm32/stm32l1/power.c (hunk 1 above)
Comment (abelNoelleC) [reply soc/arm/st_stm32/stm32l1/power.c:55]: Okay, will do. Do want it to be in a switch case similar to the one used in `pm_power_state_set()`, since that clarifies that this is required specifically for the stop mode (in case something different is required for other low-power modes)?

**Comment by abelNoelleC** [comment]
@erwango Yes, thank you. 
Although I'm still thinking about to what extend it would be required in this specific PR. It would also make sense to move it to the application layer, since (I believe) you'd want to reconfigure the clock before doing any other peripheral reconfiguration post wake-up, and the PM notifier is called before the SoC specific pm hook (see `subsys/power/power.c`). 
But that also ties into the discussion of how/when the GPIO and peripheral driver reconfiguration post wake-up will be implemented on a larger scale, and right now the `pm_resume_devices()` is called BEFORE the PM notifier (application layer) AND  SoC specific pm hook. 

So basically:
- before (peripheral&gpio) device pm implemented: clock reconfiguration & peripheral+GPIO reconfig in application layer (to ensure the clock reconfiguration can be done before other peripherals) > should provide sample?
- after (peripheral&gpio) device pm implemented: clock reconfiguration needs to be done before device pm (but how?)

But then again, this is based on the belief that the clock reconfiguration needs to be done before the other peripherals. I'm quite sure of this, but confirmation would be great.

File: soc/arm/st_stm32/stm32l1/power.c (hunk 1 above)
Comment (erwango) [reply soc/arm/st_stm32/stm32l1/power.c:55]: My point was to do it unconditionally and so in `stm32_power_init`. Since you do it when entering Low power mode and don't "undo it", then it can be done unconditionally at startup, no need to wait to enter LP mode.

Comment (abelNoelleC) [reply soc/arm/st_stm32/stm32l1/power.c:55]: Yes I understand, but I want to ensure that these configurations are applicable for all non-run modes, in case more are added. `LL_PWR_EnableUltraLowPower()` for example isn't.

Comment (erwango) [reply soc/arm/st_stm32/stm32l1/power.c:55]: The functions that are not applicable in non-run modes should be undone in `pm_power_state_exit_post_ops`.
The functions that do not need to be undone in `pm_power_state_exit_post_ops` should be put in `stm32_power_init`.

**Comment by abelNoelleC** [comment]
`LL_PWR_DisableUltraLowPower()` (PWR_CR_ULP) & `LL_PWR_SetRegulModeLP(LL_PWR_REGU_LPMODES_MAIN)` (PWR_CR_LPSDSR) now added to`pm_power_state_exit_post_ops`.

**Comment by erwango** [comment]
> Although I'm still thinking about to what extend it would be required in this specific PR. It would also make sense to move it (clock reconfiguration) to the application layer, since (I believe) you'd want to reconfigure the clock before doing any other peripheral reconfiguration post wake-up, and the PM notifier is called before the SoC specific pm hook (see subsys/power/power.c).
But that also ties into the discussion of how/when the GPIO and peripheral driver reconfiguration post wake-up will be implemented on a larger scale, and right now the pm_resume_devices() is called BEFORE the PM notifier (application layer) AND SoC specific pm hook.

I had a closer look to `pm_policy_mgr`. What we could observe is that actually the exit from sleep mode is not done directly in `pm_policy_mgr`. The first pm function executed post wake up is `pm_system_resume` (via any kind of wake up source: timer, 
 button, ...) which executes `pm_power_state_exit_post_ops` and then processing of `pm_policy_mgr` is resumed, leading to devices resumption.
So clock reconfiguration, when implemented in `pm_power_state_exit_post_ops` will happen before device sleep exit, which is fine.
As a consequence  there is no need to do specific processing on application (and by the way I think the `pm_state_notify` call before `pm_power_state_exit_post_ops` is really something that should be reconsidered). And I think this is the best option as from SW layering, application portability and scheduling perspectives, it is safer to keep this processing at SoC level.

**Comment by erwango** [comment]
CI Failures: https://github.com/zephyrproject-rtos/zephyr/issues/32344

**Comment by abelNoelleC** [comment]
rebase + add clock reconfiguration

**Comment by abelNoelleC** [comment]
To clarify: even though right now the clock reconfiguration has been implemented, I discussed it with @erwango and we have decided to keep it DNM until a LP timer has been implemented first (otherwise this PR can't be used, hence there's no real advantage to merging it already).

**Comment by abelNoelleC** [comment]
`#include <stm32_ll_utils.h>` also now included. Required for `#include <clock_control/clock_stm32_ll_common.h>`, in the case that the board uses `CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y`

**Comment by abelNoelleC** [comment]
Please remove stale label.

The PR is still relevant, and will continue working on this at some point, but we have other priorities right now.
